File: library/src/main/java/me/shkschneider/skeleton/network/GsonArrayRequest.java
package me.shkschneider.skeleton.network;

import android.support.annotation.NonNull;

import com.android.volley.NetworkResponse;
import com.android.volley.ParseError;
import com.android.volley.Response;
import com.android.volley.toolbox.HttpHeaderParser;
import com.android.volley.toolbox.JsonRequest;
import com.google.gson.Gson;
import com.google.gson.JsonArray;

import java.io.UnsupportedEncodingException;

import me.shkschneider.skeleton.data.CharsetHelper;

// <https://android.googlesource.com/platform/frameworks/volley/+/master/src/main/java/com/android/volley/toolbox/JsonArrayRequest.java>
public class GsonArrayRequest extends JsonRequest<JsonArray> {

    public GsonArrayRequest(final int method, @NonNull final String url, @NonNull final String requestBody, final Response.Listener<JsonArray> listener, final Response.ErrorListener errorListener) {
        super(method, url, requestBody, listener, errorListener);
    }

    public GsonArrayRequest(@NonNull final String url, final Response.Listener<JsonArray> listener, final Response.ErrorListener errorListener) {
        super(Method.GET, url, null, listener, errorListener);
    }

    public GsonArrayRequest(final int method, @NonNull final String url, final Response.Listener<JsonArray> listener, final Response.ErrorListener errorListener) {
        super(method, url, null, listener, errorListener);
    }

    @Override
    protected Response<JsonArray> parseNetworkResponse(@NonNull final NetworkResponse response) {
        try {
            final String string = new String(response.data, HttpHeaderParser.parseCharset(response.headers, CharsetHelper.UTF8));
            final JsonArray jsonArray = new Gson().fromJson(string, JsonArray.class);
            return Response.success(jsonArray, HttpHeaderParser.parseCacheHeaders(response));
        }
        catch (final UnsupportedEncodingException e) {
            return Response.error(new ParseError(e));
        }
    }

}

File: library/src/main/java/shkschneider/skeleton/network/GsonArrayRequest.kt
package me.shkschneider.skeleton.network

import com.android.volley.NetworkResponse
import com.android.volley.ParseError
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.HttpHeaderParser
import com.android.volley.toolbox.JsonRequest
import com.google.gson.Gson
import com.google.gson.JsonArray
import me.shkschneider.skeleton.data.CharsetHelper
import java.io.UnsupportedEncodingException
import java.nio.charset.Charset

// <https://android.googlesource.com/platform/frameworks/volley/+/master/src/main/java/com/android/volley/toolbox/JsonArrayRequest.java>
class GsonArrayRequest : JsonRequest<JsonArray> {

    constructor(method: Int, url: String, requestBody: String, listener: Response.Listener<JsonArray>, errorListener: Response.ErrorListener) : super(method, url, requestBody, listener, errorListener)

    constructor(url: String, listener: Response.Listener<JsonArray>, errorListener: Response.ErrorListener) : super(Request.Method.GET, url, null, listener, errorListener)

    constructor(method: Int, url: String, listener: Response.Listener<JsonArray>, errorListener: Response.ErrorListener) : super(method, url, null, listener, errorListener)

    override fun parseNetworkResponse(response: NetworkResponse): Response<JsonArray> {
        try {
            val string = String(response.data, Charset.forName(HttpHeaderParser.parseCharset(response.headers, CharsetHelper.UTF8)))
            val jsonArray = Gson().fromJson(string, JsonArray::class.java)
            return Response.success(jsonArray, HttpHeaderParser.parseCacheHeaders(response))
        } catch (e: UnsupportedEncodingException) {
            return Response.error(ParseError(e))
        }
    }

}
